text/0000-safe-mem-forget.md

```diff
@@ -0,0 +1,124 @@
+[known][dtor-bug1] [bugs][dtor-bug2] where `panic!` may fail to run
+destructors. Furthermore, it has [also been pointed out][drain-bug] that not
+running destructors can affect the safety of APIs like `Vec::drain_range` in
+addition to `thread::scoped`.
+
+[forget-bug]: https://github.com/rust-lang/rust/issues/24456
+[dtor-comment]: https://github.com/rust-lang/rust/issues/24292#issuecomment-93505374
+[dtor-bug1]: https://github.com/rust-lang/rust/issues/14875
+[dtor-bug2]: https://github.com/rust-lang/rust/issues/16135
+[drain-bug]: https://github.com/rust-lang/rust/issues/24292#issuecomment-93513451
+
+It has never been a guarantee of Rust that destructors for a type will run, and
+this aspect was overlooked with the `thread::scoped` API which requires that its
+destructor be run! Reconciling these two desires has lead to a good deal of
+discussion of possible mitigation strategies for various aspects of this
+problem. This strategy proposed in this RFC aims to fit uninvasively into the
+standard library to avoid large overhauls or destabilizations of APIs.
+
+# Detailed design
+
+Primarily, the `unsafe` annotation on the `mem::forget` function will be
+removed, allowing it to be called from safe Rust. This transition will be made
+possible by stating that destructors **may not run** in all circumstances (from
+both the language and library level). The standard library and the primitives it
+provides will always attempt to run destructors, but will not provide a
+guarantee that destructors will be run.
+
+It is still likely to be a footgun to call `mem::forget` as memory leaks are
+almost always undesirable, but the purpose of the `unsafe` keyword in Rust is to
+indicate **memory unsafety** instead of being a general deterrent for "should be
+avoided" APIs. Given the premise that types must be written assuming that their
+destructor may not run, it is the fault of the type in question if `mem::forget`
+would trigger memory unsafety, hence allowing `mem::forget` to be a safe
+function.
+
+Note that this modification to `mem::forget` is a breaking change due to the
+signature of the function being altered, but it is expected that most code will
+not break in practice and this would be an acceptable change to cherry-pick into
+the 1.0 release.
+
+# Drawbacks
+
+It is clearly a very nice feature of Rust to be able to rely on the fact that a
+destructor for a type is always run (e.g. the `thread::scoped` API). Admitting
+that destructors may not be run can lead to difficult API decisions later on and
+even accidental unsafety. This route, however, is the least invasive for the
+standard library and does not require radically changing types like `Rc` or
+fast-tracking bug fixes to panicking destructors.
+
+# Alternatives
+
+The main alternative this proposal is to provide the guarantee that a destructor
+for a type is always run and that it is memory unsafe to not do so. This would
```

Comment:
> what we'd be interested in preventing is Rust code being able to observe the fact of a destructor having not run when it "should have".

This is pretty easy, actually, unless you assume objects and destructors are “pure”. A nice example would be not running destructor for PID-like files that are only supposed to exist for the lifetime of process. Killing and running the same Rust code again could somewhat reliably tell you that the destructor was not run.

What I’m trying to say here: this alternative would require marking all of these prematurely-kills-process functions unsafe as well. I believe this makes the alternative unviable.